### ISSUE TITLE
[CT-717] Add Grants to BigQuery Materializations

### ISSUE THREAD

**Issue by nathaniel-may** [dbt-labs/dbt-bigquery#198]
blocked by: https://github.com/dbt-labs/dbt-core/issues/5263

Once the above work is merged into core, there are some tweaks that need to be made in the BigQuery adapter to fully expose the feature:
- [ ] add a call to apply_grants(..., revoke=True) to all custom materializations
- [ ] override `apply_grants` so that it directly makes the API call to revoke grants ([BigQuery Docs](https://cloud.google.com/bigquery/docs/samples/bigquery-revoke-dataset-access#bigquery_revoke_dataset_access-python)) instead of getting sql from `get_revoke_all_sql`.

**Comment by jtcohen6** [comment]
Following up on latest thinking in https://github.com/dbt-labs/dbt-core/issues/5263#issuecomment-1145773353, there may actually be a way to do this that's SQL-only, without any API calls. That may be slightly slower, but it's much more visible, and easier to debug.

Docs:
- https://cloud.google.com/bigquery/docs/table-access-controls
- https://cloud.google.com/bigquery/docs/information-schema-object-privileges

Defaults should work for:
- `get_grant_sql`
- `get_revoke_sql`, IFF we take the "more sophisticated" approach of calculating diffs, since BQ only supports `revoke [specific_privilege]`, not `revoke all`

We'd need a custom version of `get_show_grants_sql`, which should look like (pseudo code):
```sql
select * from {{ relation.project }}.`{{ target.location }}`.INFORMATION_SCHEMA.OBJECT_PRIVILEGES
where object_schema = '{{ relation.dataset }}' and object_name = '{{ relation.identifier }}'
```

### Example flow
```sql
grant `roles/bigquery.dataViewer` on dbt_jcohen.my_table to "user:jeremy@example.com";
```
```sql
select * from `region-us`.INFORMATION_SCHEMA.OBJECT_PRIVILEGES
where object_schema = 'dbt_jcohen' and object_name = 'my_table';
```
```json
{  "object_catalog": "dbt-test-env",  "object_schema": "dbt_jcohen",  "object_name": "my_table",  "object_type": "TABLE",  "privilege_type": "roles/bigquery.dataViewer",  "grantee": "user:jeremy@example.com"}
```
```sql
revoke `roles/bigquery.dataViewer` on dbt_jcohen.my_table from "user:jeremy@example.com";
```

The cool thing about BigQuery is that we could actually perform the show + revoke + grant statements _all in one_ dynamic "script," using ["procedural SQL"](https://cloud.google.com/bigquery/docs/reference/standard-sql/procedural-language), if we so chose. That's fairly different from the general implementation we'd need on other databases, though.

**Comment by avipaul6** [comment]
Will this also work for Authorized Views? @jtcohen6 
I can't find any SQL example of revoking just that view from the dataset without having to 
1. call the dataset, 
2. get the whole list of views and 
3. remove just that view
4. update the dataset. 

I can see that done via API calls in Python from https://cloud.google.com/bigquery/docs/samples/bigquery-revoke-dataset-access

**Comment by jtcohen6** [comment]
@avipaul6 This will not work for authorized views, for the reasons you've indicated.

dbt already supports authorized views on BigQuery today, but the syntax is pretty wonky, and differs from the standard we're looking to implement in this feature: https://docs.getdbt.com/reference/resource-configs/bigquery-configs#authorized-views

Our main constraint here by what BigQuery StandardSQL makes available. I'm definitely interested in future work that reconciles / rationalizes dbt's approach to standard grants with the dataset operations required by authorized views.
